WalletWasabi/Services/WalletService.cs

```diff
@@ -538,198 +563,7 @@ private async Task<bool> ProcessBlockAsync(Height height, Block block, IEnumerab
-						foreach (SmartCoin doubleSpentCoin in doubleSpends)
-						{
-							Coins.TryRemove(doubleSpentCoin);
-						}
-						walletRelevant = true;
-					}
-				}
-			}
-
-			var isLikelyCoinJoinOutput = false;
-			bool hasEqualOutputs = tx.Transaction.GetIndistinguishableOutputs(includeSingle: false).FirstOrDefault() != default;
-			if (hasEqualOutputs)
-			{
-				var receiveKeys = KeyManager.GetKeys(x => tx.Transaction.Outputs.Any(y => y.ScriptPubKey == x.P2wpkhScript));
-				bool allReceivedInternal = receiveKeys.All(x => x.IsInternal);
-				if (allReceivedInternal)
-				{
-					// It is likely a coinjoin if the diff between receive and sent amount is small and have at least 2 equal outputs.
-					Money spentAmount = Coins.Where(x => tx.Transaction.Inputs.Any(y => y.PrevOut.Hash == x.TransactionId && y.PrevOut.N == x.Index)).Sum(x => x.Amount);
-					Money receivedAmount = tx.Transaction.Outputs.Where(x => receiveKeys.Any(y => y.P2wpkhScript == x.ScriptPubKey)).Sum(x => x.Value);
-					bool receivedAlmostAsMuchAsSpent = spentAmount.Almost(receivedAmount, Money.Coins(0.005m));
-
-					if (receivedAlmostAsMuchAsSpent)
-					{
-						isLikelyCoinJoinOutput = true;
-					}
-				}
-			}
-
-			List<SmartCoin> spentOwnCoins = null;
-
-			for (var i = 0U; i < tx.Transaction.Outputs.Count; i++)
-			{
-				// If transaction received to any of the wallet keys:
-				var output = tx.Transaction.Outputs[i];
-				HdPubKey foundKey = KeyManager.GetKeyForScriptPubKey(output.ScriptPubKey);
-				if (foundKey != default)
-				{
-					walletRelevant = true;
-
-					foundKey.SetKeyState(KeyState.Used, KeyManager);
-					if (output.Value <= ServiceConfiguration.DustThreshold)
-					{
-						continue;
-					}
-
-					spentOwnCoins = spentOwnCoins ?? Coins.Where(x => tx.Transaction.Inputs.Any(y => y.PrevOut.Hash == x.TransactionId && y.PrevOut.N == x.Index)).ToList();
-					var anonset = tx.Transaction.GetAnonymitySet(i);
-					if (spentOwnCoins.Count != 0)
-					{
-						anonset += spentOwnCoins.Min(x => x.AnonymitySet) - 1; // Minus 1, because do not count own.
-					}
-
-					SmartCoin newCoin = new SmartCoin(txId, i, output.ScriptPubKey, output.Value, tx.Transaction.Inputs.ToTxoRefs().ToArray(), tx.Height, tx.IsRBF, anonset, isLikelyCoinJoinOutput, foundKey.Label, spenderTransactionId: null, false, pubKey: foundKey); // Do not inherit locked status from key, that's different.
-
-					if (Coins.TryAdd(newCoin))
-					{
-						TransactionCache.TryAdd(tx);
-
-						// If it's being mixed and anonset is not sufficient, then queue it.
-						if (newCoin.Unspent && ChaumianClient.HasIngredients && newCoin.AnonymitySet < ServiceConfiguration.MixUntilAnonymitySet && newCoin.IsLikelyCoinJoinOutput && ChaumianClient.State.Contains(newCoin.SpentOutputs))
-						{
-							try
-							{
-								await ChaumianClient.QueueCoinsToMixAsync(newCoin);
```

Comment:
@nopara73 in the refactored code the dependency with `ChaumianClient` was removed and replaced with an event raising as follow: 

```c#
CoinReceived?.Invoke(this, newCoin);
```

The `WalletService` is subscribed to this `CoinReceived` event  and then it queue the received coin to mix.